File: leetcode/queues/225-implement-stack-using-queues/c/my_stack.c
#include <stdlib.h>
#include <stdbool.h>
#include "my_stack.h"
#include "queue_linked_list.h"

MyStack *myStackCreate() {
  MyStack *stack = calloc(1, sizeof(MyStack));
  stack->queue = queueCreate();
  return stack;
}

void myStackFree(MyStack *stack) {
  queueDelete(stack->queue);
  free(stack);
}

void myStackPush(MyStack *stack, int x) {
  queueEnqueue(stack->queue, x);
}

int myStackPop(MyStack *stack) {
  // Dequeue & Enqueue `n-1` times to bring the last item to front
  int size = queueSize(stack->queue);

  for (int i = 0; i < size - 1; i++) {
    int item = queueDequeue(stack->queue);
    queueEnqueue(stack->queue, item);
  }

  int poppedItem = queueDequeue(stack->queue);
  return poppedItem;
}

int myStackTop(MyStack *stack) {
  // Dequeue & Enqueue `n-1` times to bring the last item to front
  int size = queueSize(stack->queue);

  for (int i = 0; i < size - 1; i++) {
    int item = queueDequeue(stack->queue);
    queueEnqueue(stack->queue, item);
  }

  int peekedItem = queuePeek(stack->queue);

  // Put the first item back to it's last position
  int item = queueDequeue(stack->queue);
  queueEnqueue(stack->queue, item);

  return peekedItem;
}

bool myStackEmpty(MyStack *stack) {
  return isQueueEmpty(stack->queue);
}

File: leetcode/queues/225-implement-stack-using-queues/ts/my-stack.ts
class MyQueue {
  #entries: number[];

  constructor() {
    this.#entries = [];
  }

  enqueue(x: number) {
    this.#entries.push(x);
  }

  dequeue(): number | undefined {
    return !this.isEmpty() ? this.#entries.shift() : undefined;
  }

  peek(): number | undefined {
    return !this.isEmpty() ? this.#entries[0] : undefined;
  }

  isEmpty() {
    return this.#entries.length === 0;
  }

  size() {
    return this.#entries.length;
  }
}

class MyStack {
  #queue: MyQueue;

  constructor() {
    this.#queue = new MyQueue();
  }

  push(x: number): void {
    this.#queue.enqueue(x);
  }

  pop(): number {
    // Dequeue & Enqueue n-1 times to bring the last item to the front
    const queueSize = this.#queue.size();

    for (let i = 0; i < queueSize - 1; i++) {
      const dequeuedItem = this.#queue.dequeue();
      dequeuedItem && (this.#queue.enqueue(dequeuedItem));
    }

    const poppedItem = this.#queue.dequeue()!;
    return poppedItem;
  }

  top(): number {
    // Dequeue & Enqueue n-1 times to bring the last item to the front
    const queueSize = this.#queue.size();

    for (let i = 0; i < queueSize - 1; i++) {
      const dequeuedItem = this.#queue.dequeue();
      dequeuedItem && (this.#queue.enqueue(dequeuedItem));
    }

    const peekedItem = this.#queue.peek()!;

    // Move the first item to it's original position
    const dequeuedItem = this.#queue.dequeue();
    dequeuedItem && (this.#queue.enqueue(dequeuedItem));

    return peekedItem;
  }

  empty(): boolean {
    return this.#queue.isEmpty();
  }
}

const stack = new MyStack();
stack.push(1);
stack.push(2);
stack.push(3);
stack.push(4);
stack.push(5);
console.log(stack.pop());
console.log(stack.pop());
console.log(stack.top());
console.log(stack.pop());
console.log(stack.top());
console.log(stack.empty());
